endpoints/endpoints.py

```diff
@@ -0,0 +1,1426 @@
+
+    return logging.getLogger(__file__)
+
+def process_roadblocks(callbacks = None, roadblock_id = None, endpoint_label = None, endpoint_deploy_timeout = None, max_roadblock_attempts = None, roadblock_password = None, new_followers = None, roadblock_messages_dir = None, roadblock_timeouts = None, max_sample_failures = None, engine_commands_dir = None, endpoint_dir = None):
+    """
+    Process the beginning and ending roadblocks associated with synchronizing a test
+
+    Args:
+        callbacks (dict): A dictionary of callbacks to endpoint specific actions.  It can define:
+                          engine-init
+                          collect-sysinfo
+                          test-start
+                          test-stop
+                          remote-cleanup
+        roadblock_id (str): The base ID to use as part of a roadblock's name
+        endpoint_label (str): The name of the calling endpoint
+        endpoint_deploy_timeout (int): The computed timeout value for endpoint engine deployment
+        max_roadblock_attempts (int): The maximum number of attempts that should be made for each roadblock
+        roadblock_password (str): The password that roadblock uses to connect to it's server
+        new_followers (list): A list of the new followers to inform the roadblock leader about
+        roadblock_messages_dir (str): The directory where roadblock messages should be stored
+        roadblock_timeouts (dict): The roadblock timeout values from rickshaw-settings
+        max_sample_failures (int): The maximum number of times a sample can fail before the iteration is considered a failure
+        engine_commands_dir (str): Directory where the engine commands can be found
+        endpoint_dir (str): The base endpoint directory for storing endpoint specific information
+
+    Globals:
+        None
+
+    Returns:
+        0
+    """
+    log.info("Starting to process roadblocks")
+
+    new_followers_msg_payload = {
+        "new-followers": new_followers
+    }
+    new_followers_msg = create_roadblock_msg("all", "all", "user-object", new_followers_msg_payload)
+
+    new_followers_msg_file = roadblock_messages_dir + "/new-followers.json"
+    log.info("Writing new followers message to %s" % (new_followers_msg_file))
+    with open(new_followers_msg_file, "w", encoding = "ascii") as new_followers_msg_file_fp:
+        new_followers_msg_file_fp.write(dump_json(new_followers_msg))
+
+    rc = do_roadblock(roadblock_id = roadblock_id,
+                      follower_id = endpoint_label,
+                      label = "endpoint-deploy-begin",
+                      timeout = endpoint_deploy_timeout,
+                      messages = new_followers_msg_file,
+                      max_attempts = max_roadblock_attempts,
+                      redis_password = roadblock_password,
+                      msgs_dir = roadblock_messages_dir)
+    if rc != 0:
```

Comment:
these blocks are identical, do you keep them separate for readability reasons ?

Reply:
Yes, the duplication is about simplicty and readability.  We actually had a loop based implementation written in Bash at one point and it was particularly confusing due to the different exceptions that were having to be made for the different tiers.

This code is essentially duplicated in 3 different places (tiers):

1. `rickshaw-run`
2. the endpoint(s)
3. the engine

Each tier has different exceptions and activities that happen in between the roadblocks.  When each one was implemented in a loop we were forced to look at 3 different loops with varying exceptions and visualizing and understanding it was very, very difficult.  So one of the first things I did was unroll the loops so that now you can search for a specific roadblock (ex. `collect-sysinfo-end`) in each location and very easily see what each "tier" of the stack is doing compared to the others.

Additionally, if we want to add additional stages you just go to the proper place in the flow and add more code (probably via copy/paste) because it is so much easier to determine what the proper place is.